### ISSUE TITLE
Request for Registry Mirror capability

### ISSUE THREAD

**Issue by ashaver** [jhwohlgemuth/tomo-cli#181]
Summary: This request involves the support of disconnected (or proxy) installs. It is a request to modify tomo-cli to take registry settings via command line arguments, project files, or global settings.

Details: In cases where disconnected installations of nodeJS exist, it is common to use a mirror of the NPM. For example, see the instructions on connecting Sonatype/Nexus ([link](https://blog.sonatype.com/using-nexus-3-as-your-repository-part-2-npm-packages)). Node has standard patterns to accept registries for installation of libraries from the mirror. For example, it is possible to install tomo-cli and surge from an insecure repository with
```
npm config set strict-ssl false
npm i  --registry https://nexus/repository/npm-all/ --global tomo-cli
npm i  --registry https://nexus/repository/npm-all/ --global surge
```

It would be wonderful if the tomo scaffolding also leveraged mirror registry settings to allow building applications when disconnected.

**Comment by jhwohlgemuth** [comment]
I would imagine that you are referring to passing arguments to the tomo API's [`install` function](https://github.com/jhwohlgemuth/tomo-cli/blob/master/src/utils/index.js#L160) which currently only executes `npm install` with options for version and dependency type?

I believe I understand why you might want this, but I do not see the merit for two reasons:
1. Adding such an ability would add more complexity to an already "robust" CLI API (use `tomo -h` to see all the available commands and options). I am slow to eschew simplicity without good cause...
2. Regarding good cause - My preferred method of handing npm configuration is to use npm configuration files. For example, I like to spin up a proxy using the [Nexus docker image](https://hub.docker.com/r/sonatype/nexus3/) and then use the command line tool, [npmrc](https://www.npmjs.com/package/npmrc), to manage my npm configurations.

Here is an example:
```bash
# create a new config, called "nexus" - one time
npmrc -c nexus
# customize config - one time, or whenever you need to
npm config set registry https://nexus/repository/npm-all/

# use config
npm install tomo-cli --global
# tomo will leverage the active npm config
tomo new app --use-react
```

**If you do not need more than one config file**, `npmrc` is overkill and you can simply edit [the configuration file at the appropriate level](https://docs.npmjs.com/configuring-npm/npmrc.html) for your needs using the "customize config" code in the above example.

***All that being said***, if using `npmrc` or editing the appropriate `.npmrc` file does not work for you, please give me more details. Assuming you have a use-case that is reasonable and cannot be mitigated by updating npm configuration files, I am not opposed to a PR that adds an `--install-options` option or something better :)

> **Note:** The `-i` flag should ignore warning messages like the "offline" warning

**Comment by ashaver** [comment]
I do not think I explained the request well enough. The registry is installed in npm and works to install tomo. 
So the below works, leveraging the registry to install (while being disconnected from the internet):
```
npm config set registry https://nexus/repository/npm-all/
npm install tomo-cli --global
```

Later when using tomo  `tomo new app --use-react` it detects that it is not connected to the internet (which it is not) and then does not install tooling. Looking at code, it appears that src/utils/index.js uses the npm package is-online to determine if items should be installed: 
```
import isOnline from 'is-online';
...
const isNotOffline = skipInstall || await isOnline();
```
The package 'is-online' checks some very specific items to determine if the execution is online (https://www.npmjs.com/package/is-online). These checks can determine internet connectivity, but if one is using a mirror, that does not equate to the inability to install packages in npm.

Anyway, I would love to be able to have tomo-cli leverage a registry when it is registered with npm, allowing applications to be built in disconnected environments.

**Comment by jhwohlgemuth** [comment]
@ashaver 

You know what? I thought maybe that's what you were referring to...but I had already written the stuff about `npmrc`...

In any case, that makes perfect sense. My initial thought is to add an `--offline` option that will disable the offline warning and allow for installing from a proxy while "offline". What do you think?

I will leave this issue open until I make a PR. I also made a Trello card for this new feature ([link](https://trello.com/c/MAdKM2qw/258-add-offline-option-to-allow-for-using-proxy-when-offline))

**Comment by jhwohlgemuth** [comment]
Actually, maybe I will just remove the online checks in the commands and always allow `npm install` to try and install...? Then you could just use `--ignore-errors` to silence the offline warning.

**Comment by jhwohlgemuth** [comment]
Too late! I [did it](https://github.com/jhwohlgemuth/tomo-cli/releases/tag/v1.6.1)

Maybe you could give it a shot and tell me if it works.

```bash
tomo new app -ri
```

**Comment by ashaver** [comment]
I grabbed the latest zip/tar of master. The behavior still seems to be to skip installing dependencies if not connected to the internet.  From console logs:

```
[root@shaver-dev1 tomo-cli-master]#  npm config set registry https://nexus/repository/npm-all/
[root@shaver-dev1 tomo-cli-master]# npm i --global tomo-cli
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
/usr/local/lib/nodejs/node-v12.18.2-linux-x64/bin/tomo -> /usr/local/lib/nodejs/node-v12.18.2-linux-x64/lib/node_modules/tomo-cli/dist/index.js

> ejs@3.1.3 postinstall /usr/local/lib/nodejs/node-v12.18.2-linux-x64/lib/node_modules/tomo-cli/node_modules/ejs
> node --harmony ./postinstall.js

Thank you for installing EJS: built with the Jake JavaScript build tool (https://jakejs.com/)


> core-js@3.6.5 postinstall /usr/local/lib/nodejs/node-v12.18.2-linux-x64/lib/node_modules/tomo-cli/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"

Thank you for using core-js ( https://github.com/zloirock/core-js ) for polyfilling JavaScript standard library!

The project needs your help! Please consider supporting of core-js on Open Collective or Patreon:
> https://opencollective.com/core-js
> https://www.patreon.com/zloirock

Also, the author of core-js ( https://github.com/zloirock ) is looking for a good job -)

+ tomo-cli@1.6.1
added 531 packages from 302 contributors in 13.408s
[root@shaver-dev1 tomo-cli-master]# tomo new app -ri

 ▶ ▶ ▶  00:00:16

 ┌───────────────────────────┐
 │ (⌒_⌒;) This is awkward... │
 └───────────────────────────┘
    ↳ ...you appear to be offline
    ↳ Please connect to the internet and try again

      No dependencies will be installed


 ╭─────────╮
 │ new app │
 ╰─────────╯
    ↳ Finished 3 of 20 tasks (3 completed, 10 skipped)

```

**Comment by jhwohlgemuth** [comment]
That is what I get for thinking I did not need to test it... Thank you for trying it out. I will make another attempt and test it out before pushing it. I forgot how I wired in online support into various parts of the app.

This particular functionality is mildly annoying to test because of the initial conditions required...

**Comment by jhwohlgemuth** [comment]
# Quick Questions for you:
- I noticed that your output from above was not from a finished build. Did you try to let it finish? Does it ever finish, or does it hang?
- What operating system are you using? (I am guessing Windows? This might be a tomo + Windows + offline issue...)

# I was unable to reproduce this issue on Linux (Ubuntu)
I took the following steps:
1. Create Nexus repository using [nexus3 docker image](https://hub.docker.com/r/sonatype/nexus3/)
```bash
    docker run -d -p 8081:8081 --name nexus sonatype/nexus3
```
2. Start nexus serve - `docker start nexus`
3. Create npm proxy and npm group within nexus server
4. Point npm registry at nexus npm repo URL
5. Run `tomo new app -ri`
6. Verify app works with HMR and that nexus repo has stuff in it
7. Turn off internet and verify no connection via ping and trying to open web pages
8. Repeat steps 5-6

## **Results**:
![tomo-offline](https://user-images.githubusercontent.com/6383605/86981572-8a49f780-c14c-11ea-9198-bb822f9170da.PNG)

Aside from showing an "offline warning" that said no dependencies will be installed, it all worked quite well, and about twice as fast as when used without a local proxy.

> Note: I will be editing the "offline" warning and making sure it does not show up when `-i` or `--ignore-warnings` are used.

**Comment by jhwohlgemuth** [comment]
@ashaver 

As another possible difference in our environments, I would note that I not behind an overly restrictive firewall.

`tomo` may just be eating npm errors caused by restrictive network settings. In the past, I have also tried using `npm config set ca ""` in order to function behind strict corporate firewalls. This might apply to your situation. You might say "well, I installed tomo-cli and surge just fine". While that is true, my experience has been that *most* packages might install fine, but it only takes one to break it all...

I will try to reproduce this issue on Windows within the next few days.
